src/Microsoft.AspNetCore.Cors/Infrastructure/CorsMiddleware.cs

```diff
@@ -106,26 +106,38 @@ public class CorsMiddleware
                 var corsPolicy = _policy ?? await _corsPolicyProvider?.GetPolicyAsync(context, _corsPolicyName);
                 if (corsPolicy != null)
                 {
-                    var corsResult = _corsService.EvaluatePolicy(context, corsPolicy);
-                    _corsService.ApplyResult(corsResult, context.Response);
-
                     var accessControlRequestMethod =
                         context.Request.Headers[CorsConstants.AccessControlRequestMethod];
                     if (string.Equals(
                             context.Request.Method,
                             CorsConstants.PreflightHttpMethod,
                             StringComparison.OrdinalIgnoreCase) &&
-                            !StringValues.IsNullOrEmpty(accessControlRequestMethod))
+                        !StringValues.IsNullOrEmpty(accessControlRequestMethod))
                     {
+                        ApplyCorsHeaders(context, corsPolicy);
+
                         // Since there is a policy which was identified,
                         // always respond to preflight requests.
                         context.Response.StatusCode = StatusCodes.Status204NoContent;
                         return;
                     }
+                    else
+                    {
+                        context.Response.OnStarting(_ => {
+                            ApplyCorsHeaders(context, corsPolicy);
```

Comment:
Catch all exceptions in this callback. An exception on this call path would bypass all exception pages and result in a blank 500.

Reply:
This is middleware which means it can be used without mvc. Furthermore exceptions on this path already result in a blank 500 except it won’t have CORS headers.

You should use the exception handler middleware to provide consistent, last chance errors to the client. This will also help in cases were mvc is used with other serving middleware.

Reply:
It's not practical for the exception handling middleware to handle exceptions in OnStarting, those callbacks are invoked directly by the server.

Reply:
I’m sorry I think you’ve misunderstood (or I’m missing something). There is no exception handling in OnStarting. 

This change is primarily to ensure CORS middleware works with ‘UseExceptionHandler’ but applies to all similar cases.

The flow goes like this:
1. An exception is uncaught in an API controller.
2. Exception is handled by global exception handler.
3. CORS headers are added to error details can be used by the client.

Is there anything else I can clarify?

Reply:
@Tratcher or did you mean wrapping `ApplyCorsHeaders` in a `try catch` in case of an error in the callback?
```
context.Response.OnStarting(_ => {
    ApplyCorsHeaders(context, corsPolicy);
```

Reply:
Right. We want to avoid throwing exceptions in the OnStarting code path as it has a worse error handling experience. These wouldn't be app exceptions, but exceptions from within CORS itself. I'm not aware of any, but we'd rather not risk it.